src/hash.cr

```diff
@@ -893,17 +896,73 @@ class Hash(K, V)
 
   # Computes the hash of a key.
   private def key_hash(key)
-    hash = key.hash.to_u32!
+    if @compare_by_identity && key.responds_to?(:object_id)
+      hash = key.object_id.hash.to_u32!
+    else
+      hash = key.hash.to_u32!
+    end
     hash == 0 ? UInt32::MAX : hash
   end
 
+  private def entry_matches?(entry, hash, key)
+    # Tiny optimization: for these primitive types it's faster to just
+    # compare the key instead of comparing the hash and the key.
+    # We still have to skip hashes with value 0 (means deleted).
+    {% if K == Bool ||
+            K == Char ||
+            K == Symbol ||
+            K < Int::Primitive ||
+            K < Float::Primitive ||
+            K < Enum %}
+      entry.key == key && entry.hash != 0_u32
+    {% else %}
+      entry.hash == hash && entry_matches?(entry, key)
+    {% end %}
+  end
+
+  private def entry_matches?(entry, key)
+    entry_key = entry.key
+    if @compare_by_identity && entry_key.responds_to?(:object_id)
```

Comment:
What if `@compare_by_identity` is true and `entry` responds to `object_id` but `entry_key` doesn't? Shouldn't that also return false even if `entry_key == key`?

Reply:
I don't know. If `entry_key` is an int you still want ints to compare the same way.

I also thought about raising a compile-time error when using `compare_by_identity` on a hash where keys don't respond to `object_id`, but it involved a chunk of macro code. Would something like that be better?

In general, I wouldn't worry too much about this. If you use `compare_by_identity` you generally know what you are doing and you'll be already working with reference types, or reference-like types. But we can try adding the macro check.

Reply:
I'm not sure what's the best semantics here. But I'd try to be consistent.

Currently, when comparing by identity the behaviour for differing answers to `responds_to?(:object_id)` is inconsistent:
* when `entry_key.responds_to?(:object_id) && !key.responds_to?(:object_id)` the result is `false`
* when  `!entry_key.responds_to?(:object_id) && key.responds_to?(:object_id)` the result is `entry_key == key`

Reply:
I'm not sure what's the best semantics here. But I'd try to be consistent.

Currently, when comparing by identity the behaviour for differing answers to `responds_to?(:object_id)` is inconsistent:
* when `entry_key.responds_to?(:object_id) && !key.responds_to?(:object_id)` the result is `false`
* when  `!entry_key.responds_to?(:object_id) && key.responds_to?(:object_id)` the result is `entry_key == key`

Reply:
Sounds good, I'll at least make it consistent

Reply:
Done! Let me know what you think.